vioscsi/vioscsi.c

```diff
@@ -802,6 +803,7 @@ ENTER_FN();
                     adaptExt->perfFlags |= STOR_PERF_INTERRUPT_MESSAGE_RANGES;
                     perfData.FirstRedirectionMessageNumber = 3;
                     perfData.LastRedirectionMessageNumber = perfData.FirstRedirectionMessageNumber + adaptExt->num_queues - 1;
+                    ASSERT(perfData.LastRedirectionMessageNumber < adaptExt->num_affinity);
```

Comment:
If something is wrong, revert to using single queue (failing VioScsiHwInitialize probably not a good idea, if the driver still can function). Don't forget to log the failure.

Reply:
What I mean, is that ASSERT in "Debug" configuration of the driver (which we don't build) will crash the driver, and in "Release" configuration it just don't do anything. And crashing production driver in case were driver still can function, but maybe with not with the best performance is really not path we want to take.

Reply:
@YanVugenfirer 
Just a small comment here.
Technically, if something went wrong in the Release build the following call 
status = StorPortInitializePerfOpts(DeviceExtension, FALSE, &perfData);
https://github.com/virtio-win/kvm-guest-drivers-windows/blob/master/vioscsi/vioscsi.c#L828
will return STOR_STATUS_UNSUCCESSFUL making the driver falling  back to no performance 
optimizations at all. 
Cheers,
Vadim.

Reply:
> What I mean, is that ASSERT in "Debug" configuration of the driver (which we don't build) will crash the driver, and in "Release" configuration it just don't do anything. And crashing production driver in case were driver still can function, but maybe with not with the best performance is really not path we want to take.

Hi Yan, thanks for reviewing the patch. The ASSERT added here is intended to catch any buffer overrun in the future (eg, someone changes the way FirstRedirectionMessageNumber and LastRedirectionMessageNumber is set, which causes a regression). If LastRedirectionMessageNumber >= num_affinity, the kernel will write beyond the end of this pool allocation and it will be a bug not a legitimate case we want to handle.
I agree that ASSERT does nothing in the release build. I looked up NT_ASSERT as well but that one is also a no-op in release build. Existing code uses "ASSERT" in multiple places so I thought it would be low risk to follow that pattern.
The vioscsi path doesn't have this problem but a similar path in viostor does cause that issue. It took me a while to figure it out and I believe that if such an assert is there in the first place, that bug can be caught during development and testing phase.